src/main/protobuf/bazel_output_service.proto

```diff
@@ -0,0 +1,326 @@
+option java_outer_classname = "BazelOutputServiceProto";
+option go_package = "bazeloutputservice";
+
+// The Bazel Output Service may be used by users of the Remote Execution API to
+// construct a directory on the local system that contains all output files of a
+// build.
+//
+// Primitive implementations of this API may simply download files from the
+// Content Addressable Storage (CAS) and store them at their designated
+// location. Complex implementations may use a pseudo file system (e.g., FUSE)
+// to support deduplication, lazy loading and snapshotting.
+//
+// Details:
+// https://github.com/bazelbuild/proposals/blob/master/designs/2021-02-09-remote-output-service.md
+// https://groups.google.com/g/remote-execution-apis/c/qOSWWwBLPzo
+// https://groups.google.com/g/bazel-dev/c/lKzENsNd1Do
+// https://docs.google.com/document/d/1W6Tqq8cndssnDI0yzFSoj95oezRKcIhU57nwLHaN1qk/edit
+service BazelOutputService {
+  // Clean all data associated with a single output path, so that the
+  // next invocation of StartBuild() yields an empty output path. This
+  // may be implemented in a way that's faster than removing all of the
+  // files from the file system manually.
+  rpc Clean(CleanRequest) returns (CleanResponse);
+
+  // Signal that a new build is about to start.
+  //
+  // Bazel uses this call to obtain a directory where outputs of the build may
+  // be stored, called the output path. Based on the parameters provided, server
+  // may provide an empty output path, or one that has contents from a previous
+  // build of the same workspace.
+  //
+  // In case the output path contains data from a previous build, server is
+  // responsible for calling ContentAddressableStorage.FindMissingBlobs() for
+  // all of the objects that are stored remotely. This ensures that these
+  // objects don't disappear from the Content Addressable Storage while the
+  // build is running. Any files that are absent must be removed from the output
+  // path and reported through InitialOutputPathContents.modified_paths.
+  rpc StartBuild(StartBuildRequest) returns (StartBuildResponse);
+
+  // Stage build artifacts at the given paths with digests that are known to the
+  // Content Addressable Storage. Contents of these files can be lazily staged.
+  //
+  // If any of the digests or paths are invalid an error is returned without
+  // modifying anything. If some digests are unavailable, the corresponding
+  // StageArtifactsResponse.Response.unavailable flag should be set.
+  rpc StageArtifacts(StageArtifactsRequest) returns (StageArtifactsResponse);
+
+  // Notifies server that a set of files will no longer be modified by Bazel or
+  // local build actions. Server should unset "dirty" bit for these files if it
+  // tracks file modifications under output path. If these paths are modified
+  // via FUSE operations later, they should be reported back to Bazel in the
+  // next InitialOutputPathContents.modified_paths for the same workspace.
```

Comment:
Maybe we should document the interaction with StageArtifacts? As in, Bazel does not need to call this function for any artifacts that it creates through StageArtifacts. This method only needs to be called for locally created artifacts.

Reply:
We intend to call this function for all declared action outputs which may include outputs created with `StageArtifacts`. Anyway, I don't think it matters, servers can ignore paths that they know are immutable.

Reply:
On second thought: the deletion case for test attempts doesn't really apply here, because it occurs during action execution, before finalization. But deletion through direct filesystem manipulation still matters.